File: web/src/net/endpoint.ts
/**
 * Where the Mac is, and whether this page is allowed to reach it.
 *
 * The phone had one answer to this: `http://<address>:<port>`. A browser has
 * three, and one of them is a refusal the user has to be told about precisely
 * rather than left to read as "the Mac is down".
 *
 *  1. This bundle served *by the host itself* at `http://<mac>:8787/`. Same
 *     origin as the API, no scheme mismatch, works over Tailscale and the LAN.
 *     This is the local path and it is the one that needs no configuration.
 *
 *  2. This bundle served from GitHub Pages, i.e. `https://…`. A page on `https`
 *     may not open `http://` or `ws://` — the browser blocks it as mixed content
 *     before a packet leaves. So from Pages the only reachable address is an
 *     `https` one, which in practice means the Cloudflare Tunnel the host can
 *     start (Settings → Relay over internet).
 *
 *  3. `npm run dev` on `http://localhost:5273`, talking to the Mac on `:8787`.
 *
 * `describe()` below is what turns case 2 into a sentence with an instruction in
 * it instead of a timeout.
 */

export interface Endpoint {
  /** Canonical origin, no trailing slash: `http://192.168.1.24:8787`. */
  origin: string
  /** Hostname alone, for display and for the paired-host record. */
  host: string
  /** The port actually in play, including the 80/443 a bare https URL implies. */
  port: number
  secure: boolean
}

export class EndpointError extends Error {}

/**
 * Parses what the user typed. Accepts a bare address, an address with a port, or
 * a full URL — because the tunnel hostname is something they will paste, and
 * demanding it be split into host and port is a needless trap.
 */
export function parseEndpoint(input: string, fallbackPort = 8787): Endpoint {
  const trimmed = input.trim().replace(/\/+$/, '')
  if (!trimmed) throw new EndpointError('Enter the host’s address first.')

  const hasScheme = /^[a-z][a-z0-9+.-]*:\/\//i.test(trimmed)
  // A bare `host:port` is not a URL, and `new URL('mac:8787')` parses `mac:` as
  // a scheme. Give it one it cannot mistake.
  const candidate = hasScheme ? trimmed : `http://${trimmed}`

  let url: URL
  try {
    url = new URL(candidate)
  } catch {
    throw new EndpointError(`Not a usable address: ${input}`)
  }

  if (url.protocol !== 'http:' && url.protocol !== 'https:') {
    throw new EndpointError(`VibeWire speaks HTTP, not ${url.protocol.replace(':', '')}.`)
  }
  if (!url.hostname) throw new EndpointError(`Not a usable address: ${input}`)

  const secure = url.protocol === 'https:'
  // An explicit port wins; a bare `https://tunnel` means 443 and must not have
  // 8787 bolted onto it.
  const port = url.port
    ? Number(url.port)
    : hasScheme
      ? secure
        ? 443
        : 80
      : fallbackPort

  const origin =
    (secure ? 'https://' : 'http://') +
    url.hostname +
    (port === (secure ? 443 : 80) ? '' : `:${port}`)

  return { origin, host: url.hostname, port, secure }
}

/** `ws://` or `wss://` for the same origin. */
export function socketOrigin(endpoint: Endpoint): string {
  return endpoint.origin.replace(/^http/, 'ws')
}

/**
 * True when the browser will refuse this endpoint outright.
 *
 * The mixed-content rule has one exception that matters here: a loopback address
 * counts as potentially trustworthy in Chromium, so `http://localhost:8787` from
 * an `https` page works there — but not in Safari. Treating loopback as blocked
 * would be wrong on the browser most likely to be used for development; treating
 * it as fine would be wrong on Safari. It is reported as `warn`, with the
 * sentence saying which is which, rather than pretending to a certainty nobody
 * has.
 */
export type Reachability = 'ok' | 'blocked' | 'warn'

const LOOPBACK = /^(localhost|127(?:\.\d+){3}|\[?::1\]?)$/i

export function reachability(endpoint: Endpoint): Reachability {
  if (endpoint.secure) return 'ok'
  if (location.protocol !== 'https:') return 'ok'
  if (LOOPBACK.test(endpoint.host)) return 'warn'
  return 'blocked'
}

/**
 * The sentence to show. Names the failure precisely — a blocked scheme, a
 * loopback special case, and a Mac that simply is not answering are three
 * different states, and collapsing them into one message is the defect this
 * product has already paid for once.
 */
export function describe(endpoint: Endpoint): string | null {
  switch (reachability(endpoint)) {
    case 'blocked':
      return (
        `This page is served over HTTPS, so the browser will not open a plain ` +
        `http:// connection to ${endpoint.host}. Two ways round it: turn on ` +
        `Settings → Relay over internet on the host and use the https:// tunnel ` +
        `address it prints, or open this client from the host itself at ` +
        `http://${endpoint.host}:${endpoint.port}/.`
      )
    case 'warn':
      return (
        `Loopback over HTTPS works in Chrome and Edge and is refused by Safari. ` +
        `If nothing answers here, that is why.`
      )
    default:
      return null
  }
}

/**
 * The address to offer before the user has typed anything.
 *
 * Served by the host, this page's own origin *is* the Mac — so the field is
 * pre-filled and pairing is one code away. Served from Pages, there is nothing
 * to guess and the field stays empty rather than suggesting a wrong answer.
 */
export function suggestedAddress(): string {
  const port = location.port ? Number(location.port) : location.protocol === 'https:' ? 443 : 80
  // The dev server is not the host, and neither is a Pages origin.
  if (port === 5273 || /github\.io$/i.test(location.hostname)) return ''
  if (!location.hostname || location.protocol === 'file:') return ''
  return location.origin
}

/** True when this bundle is being served by the Mac it talks to. */
export function isHostServed(): boolean {
  const port = location.port ? Number(location.port) : 0
  return port === 8787
}

/**
 * The same parse, for strings that arrived over the wire rather than off a
 * keyboard.
 *
 * A candidate list is not worth failing a connection over: the Mac reporting one
 * address this build cannot make sense of is not a reason to drop the two it
 * can.
 */
export function lenientEndpoint(input: string): Endpoint | null {
  try {
    return parseEndpoint(input)
  } catch {
    return null
  }
}

/**
 * Candidate origins in the order they should be dialled, with duplicates and
 * unparseable entries dropped.
 *
 * Order is the whole point and it is the Mac's judgement, not this page's: the
 * tailnet address before the tunnel because one is a direct route and the other
 * is a round trip through Cloudflare. This only guarantees that the order given
 * survives and that nothing is dialled twice.
 */
export function normaliseOrigins(origins: readonly string[]): string[] {
  const seen = new Set<string>()
  const ordered: string[] = []
  for (const entry of origins) {
    const endpoint = lenientEndpoint(entry)
    if (!endpoint || seen.has(endpoint.origin)) continue
    seen.add(endpoint.origin)
    ordered.push(endpoint.origin)
  }
  return ordered
}

/**
 * Whether this page is allowed to open this origin at all.
 *
 * The one place the browser differs from the phone in a way that cannot be
 * papered over. The phone holds three addresses and may dial any of them; a page
 * on `https` may dial only the `https` one, and dialling the others is not a
 * failed connection but a `SecurityError` thrown before a packet leaves. A
 * blocked origin left in the rotation costs a real attempt and a real backoff
 * for an outcome that was decided in advance, so it is dropped here instead.
 *
 * Loopback is kept: `reachability` reports it as `warn` rather than `blocked`
 * because Chromium allows it and Safari does not, and a dial that might work is
 * worth one attempt.
 */
export function dialable(origin: string): boolean {
  const endpoint = lenientEndpoint(origin)
  return endpoint != null && reachability(endpoint) !== 'blocked'
}

File: ios/VibeWire/Net/Endpoint.swift
import Foundation

/// Where the Mac is, parsed once so every caller agrees about it.
///
/// This phone used to store an address and a port and build `http://…` and
/// `ws://…` from them by hand, which quietly decided that the Mac is always
/// reachable over plain HTTP on a numbered port. That is true of the LAN address
/// and of the tailnet one. It is not true of the Cloudflare Tunnel, which is
/// `https` on 443 and is the only address that works from cellular when the
/// phone is not on the tailnet — so the app could not use the one path that was
/// built for being away from the Mac.
///
/// The web client already had this type (`web/src/net/endpoint.ts`) and this is
/// deliberately the same shape, down to the defaulting rules: a bare `mac:8787`
/// means HTTP, an explicit `https://…` means 443 unless it says otherwise, and
/// the canonical origin never carries a port that its scheme already implies.
struct Endpoint: Equatable, Sendable {
    /// Canonical origin, no trailing slash: `http://192.168.1.24:8787`.
    let origin: String
    /// Hostname alone, for display and for the "the Mac moved" field.
    let host: String
    /// The port actually in play, including the 443 a bare `https://` implies.
    let port: Int
    let secure: Bool

    var baseURL: URL? { URL(string: origin) }

    /// `ws://` or `wss://` for the same origin. A `wss` socket is what makes the
    /// tunnel usable: an `https` endpoint refuses a plain `ws` upgrade.
    var socketURL: URL? {
        guard let scheme = origin.range(of: "://") else { return nil }
        return URL(string: "ws" + (secure ? "s" : "") + origin[scheme.lowerBound...] + "/v1/socket")
    }

    /// What to show a reader. The scheme and the default port are noise on the
    /// common case and load-bearing on the tunnel, so they are printed only when
    /// they are not the ordinary answer.
    var display: String {
        secure || port != 8787 ? origin : host
    }
}

enum EndpointError: Error, LocalizedError {
    case unusable(String)
    case unsupportedScheme(String)

    var errorDescription: String? {
        switch self {
        case .unusable(let input): return "Not a usable address: \(input)"
        case .unsupportedScheme(let scheme):
            return "VibeWire speaks HTTP, not \(scheme)."
        }
    }
}

extension Endpoint {
    /// Parses what was typed, scanned, or stored. Accepts a bare address, an
    /// address with a port, or a full URL — the tunnel hostname is something
    /// that gets pasted, and demanding it be split into a host and a port is a
    /// trap with no upside.
    static func parse(_ input: String, fallbackPort: Int = 8787) throws -> Endpoint {
        var trimmed = input.trimmingCharacters(in: .whitespacesAndNewlines)
        while trimmed.hasSuffix("/") { trimmed.removeLast() }
        guard !trimmed.isEmpty else { throw EndpointError.unusable(input) }

        // A bare `mac:8787` is not a URL — `URLComponents` reads `mac:` as the
        // scheme — so anything without one is given a scheme it cannot mistake.
        let hasScheme = trimmed.range(
            of: "^[a-zA-Z][a-zA-Z0-9+.-]*://",
            options: .regularExpression
        ) != nil
        let candidate = hasScheme ? trimmed : "http://\(trimmed)"

        guard let components = URLComponents(string: candidate),
              let host = components.host, !host.isEmpty
        else { throw EndpointError.unusable(input) }

        let scheme = (components.scheme ?? "http").lowercased()
        guard scheme == "http" || scheme == "https" else {
            throw EndpointError.unsupportedScheme(scheme)
        }

        let secure = scheme == "https"
        // An explicit port wins. A bare `https://tunnel` means 443 and must not
        // have 8787 bolted onto it; a bare `192.168.1.24` with no scheme at all
        // means the port the pairing screen is showing.
        let port = components.port ?? (hasScheme ? (secure ? 443 : 80) : fallbackPort)
        let implied = secure ? 443 : 80
        let origin = "\(scheme)://\(host)" + (port == implied ? "" : ":\(port)")

        return Endpoint(origin: origin, host: host, port: port, secure: secure)
    }

    /// The same parse, for strings that came off the wire rather than off a
    /// keyboard — a candidate list is not worth failing a connection over.
    static func lenient(_ input: String) -> Endpoint? {
        try? parse(input)
    }

    /// Candidate origins in the order they should be dialled, with duplicates
    /// and unparseable entries dropped.
    ///
    /// Order is the whole point: the tunnel is put first by the host because it
    /// is the one address that answers from anywhere, and the LAN address last
    /// because it is the one that answers fastest when it answers at all. What
    /// this function guarantees is only that the list stays in the order it was
    /// given and that nothing is tried twice.
    static func normalise(_ origins: [String]) -> [String] {
        var seen = Set<String>()
        var ordered: [String] = []
        for entry in origins {
            guard let endpoint = lenient(entry), !seen.contains(endpoint.origin) else { continue }
            seen.insert(endpoint.origin)
            ordered.append(endpoint.origin)
        }
        return ordered
    }
}

/// What the pairing QR carries, from either generation of host.
///
/// A current host writes `origin` and `alt` — full origins, the one that works
/// from anywhere first. Every host before this one wrote `host` and `port`, an
/// address on a numbered port and nothing else. Both are read here, in one
/// place, so the scanner and the cold-launch URL handler cannot disagree about
/// what a link means.
struct PairingLink {
    let origin: String?
    let alternatesField: String?
    let host: String?
    let portField: String?

    init(origin: String?, alternates: String?, host: String?, port: String?) {
        self.origin = origin.flatMap { $0.isEmpty ? nil : $0 }
        self.alternatesField = alternates
        self.host = host.flatMap { $0.isEmpty ? nil : $0 }
        self.portField = port
    }

    var port: Int { portField.flatMap(Int.init) ?? 8787 }

    /// The address to try first.
    var address: String? { origin ?? host }

    /// Everything else worth trying, in the order the Mac gave them, with the
    /// legacy `host`/`port` pair kept on the end — it is usually the tailnet or
    /// LAN address, and it is the fastest of the three when the phone is on it.
    var alternates: [String] {
        var offered = (alternatesField ?? "")
            .split(separator: ",")
            .map { $0.trimmingCharacters(in: .whitespaces) }
            .filter { !$0.isEmpty }
        if let host, origin != nil {
            offered.append("http://\(host):\(port)")
        }
        return Endpoint.normalise(offered)
    }
}
